#perl! -w

#takes a gff formatted file generated by Kang Du and reformats it work with DESeq2
#make sure to subset the gff file to only include transcript information before using:
#example:
#awk -v OFS='\t' '$3="transcript"' xvar-COAC-24-XI-21-M_v2023.2.gff | grep 'symbol=' > xvar-COAC-24-XI-21-M_v2023.2.tx.transcripts.gtf

if(@ARGV<1){
    print "perl modify_kang_gff_to_DESeq2_v2.pl gff_transcript_file\n"; exit;
}

my $infile=shift(@ARGV); chomp $infile;
open IN, $infile or die "cannot open $infile\n";

my $outfile="$infile".".reformat";
open OUT, ">$outfile";

while(my $line=<IN>){

    chomp $line;
    my @elements=split(/\t/,$line);
    my $focal=$elements[8];

    my $rm_spaces="$elements[8]"."$elements[9]"."$elements[10]"."$elements[11]"."$elements[12]"."$elements[13]"."$elements[14]"."$elements[15]"."$elements[16]";
    $rm_spaces =~ s/\t/s+/g;

    my $modified_line="$elements[0]"."\t"."$elements[1]"."\t"."$elements[2]"."\t"."$elements[3]"."\t"."$elements[4]"."\t"."$elements[5]"."\t"."$elements[6]"."\t"."$elements[7]"."\t"."$rm_spaces";

    my @deconstruct=split(/;/,$focal);
    my $gene_id=$deconstruct[0];
    
    my @tmp=split(/=/,$gene_id);
    my $gene=$tmp[1];
   # print "$gene\n";

    my $replace="$gene_id".";"."transcript_id="."$gene".".t1";

    $modified_line=~ s/$gene_id/$replace/g;
    $modified_line=~ s/ID=/gene_id=/g;

    print OUT "$modified_line\n";

}
