#perl! -w

###takes as input a list of sorted deduplicated bam files generated by "submit_bwa_jobs_list.pl" 
#ex:
#file1.sorted.dedup.bam
#file2.sorted.dedup.bam

###The slurm submission file should contain everything you need to include in your slurm batch command except for the commands to run the job, those will be added by this script

###also give the absolute path to GATK, leave blank if globally installed

###give the genome assembly used for mapping, make sure you have already generate a GATK sequence dictionary (.dict file)

###list of targets to run GATK's vcf on, *MUST* have the file extension .list and contain one scaffold name per line

if(@ARGV<5){
    print "perl submit_gatk-indel-hc_jobs_list.pl list_of_sorted_bam_files slurm_submission_file_to_use absolute_path_to_GATK_jars genome_assembly_used vcf_targets_list"
}#print usage

my $infile=shift(@ARGV); chomp $infile;
open IN, $infile or die "cannot open bam list\n";

my $slurm_file=shift(@ARGV); chomp $slurm_file;

my $gatk_path=shift(@ARGV); chomp $gatk_path;

my $genome=shift(@ARGV); chomp $genome;

my $targets_list=shift(@ARGV); chomp $targets_list;

my $counter=0;
while(my $line=<IN>){
    chomp $line;
    $counter=$counter+1;
    my $file1=$line;

    my $command=qx(cat $slurm_file); chomp $command;

    my $dedup="$file1";

    my $targets="$file1".".list";

    my $string1="java -jar "."$gatk_path"."/GenomeAnalysisTK.jar -T RealignerTargetCreator -R $genome -I $dedup -o $targets";

    my $realigned=$dedup;
    $realigned=~ s/sorted.dedup/sorted.dedup.realigned/g;
    print "realigned $realigned\n";

    my $string2="java -jar "."$gatk_path"."/GenomeAnalysisTK.jar -T IndelRealigner -R $genome -I $dedup -targetIntervals $targets -o $realigned";

    my $raw="$file1"."rawvariants.g.vcf";

    my $string3="java -jar "."$gatk_path"."/GenomeAnalysisTK.jar -T HaplotypeCaller -R $genome -I $realigned --genotyping_mode DISCOVERY -L $targets_list -stand_emit_conf 10 -stand_call_conf 30 -ERC GVCF -o $raw";

    my $gvcf="$file1".".g.vcf";

    my $string4="java -jar "."$gatk_path"."/GenomeAnalysisTK.jar -T GenotypeGVCFs -R $genome --variant $raw --sample_ploidy 2 --max_alternate_alleles 4 --includeNonVariantSites --standard_min_confidence_threshold_for_calling 30 -o $gvcf";

    open OUT, ">run_gatk-indel-hc_shell_"."$counter".".sh";
    print OUT "$command\n$string1\n$string2\n$string3\n$string4\n";

    my $script="run_gatk-indel-hc_shell_"."$counter".".sh";
    system("sbatch $script");

}
